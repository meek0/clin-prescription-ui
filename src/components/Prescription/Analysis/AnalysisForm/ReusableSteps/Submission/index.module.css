.submissionForm span[class$=-item-label] {
  width: 30%;
}
@media screen and (min-width: 1450px) {
  .submissionForm span[class$=-item-label] {
    width: 25%;
  }
}

.supervisorCommentWrapper {
  margin-left: 1px;
  margin-right: 1px;
}
.supervisorCommentWrapper :global(.ant-row) {
  flex-direction: column;
}

.reviewLabel {
  margin-top: 48px;
  margin-bottom: 12px;
}
