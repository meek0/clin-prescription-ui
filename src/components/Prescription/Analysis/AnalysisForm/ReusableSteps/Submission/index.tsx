/* eslint-disable no-console */
import { useEffect, useState } from 'react';
import intl from 'react-intl-universal';
import { FormOutlined } from '@ant-design/icons';
import Collapse, { CollapsePanel } from '@ferlab/ui/core/components/Collapse';
import ProLabel from '@ferlab/ui/core/components/ProLabel';
import { Descriptions, Form, Input, Select, Tag } from 'antd';
import { DefaultOptionType } from 'antd/lib/select';
import {
  findPractitionerRoleByOrganization,
  isPractitionerResident,
} from 'api/fhir/practitionerHelper';
import { HybridApi } from 'api/hybrid';
import { HybridAnalysis } from 'api/hybrid/models';

import AnalysisForm from 'components/Prescription/Analysis/AnalysisForm';
import {
  defaultCollapseProps,
  defaultFormItemsRules,
  STEPS_ID,
} from 'components/Prescription/Analysis/AnalysisForm/ReusableSteps/constant';
import { SubmissionStepMapping } from 'components/Prescription/Analysis/stepMapping';
import { getNamePath, setInitialValues } from 'components/Prescription/utils/form';
import { IGetNamePathParams } from 'components/Prescription/utils/type';
import { useAppDispatch } from 'store';
import { useGlobals } from 'store/global';
import { usePrescriptionForm } from 'store/prescription';
import { prescriptionFormActions } from 'store/prescription/slice';
import { useUser } from 'store/user';

import styles from './index.module.css';

const Submission = () => {
  const FORM_NAME = STEPS_ID.SUBMISSION;
  const { user } = useUser();
  const [form] = Form.useForm();
  const dispatch = useAppDispatch();
  const { getAnalysisNameByCode } = useGlobals();
  const { analysisFormData, config, currentStep, analysisType } = usePrescriptionForm();
  const [supervisors, setSupervisors] = useState<DefaultOptionType[]>([]);
  const [autoCompleteDropdownIsOpen, setAutoCompleteDropdownIsOpen] = useState(false);

  const getName = (...key: IGetNamePathParams) => getNamePath(FORM_NAME, key);

  useEffect(() => {
    if (analysisFormData.analysis.resident_supervisor_id) {
      HybridApi.searchSupervisors({
        organizationId: analysisFormData.proband?.organization_id!,
        prefix: analysisFormData.analysis.resident_supervisor_id,
      })
        .then(({ data }) => {
          form.setFieldValue(
            getName('resident_supervisor_id' satisfies keyof HybridAnalysis),
            data?.supervisors[0].name || analysisFormData.analysis.resident_supervisor_id,
          );
        })
        .catch(() => {
          form.setFieldValue(
            getName('resident_supervisor_id' satisfies keyof HybridAnalysis),
            analysisFormData.analysis.resident_supervisor_id,
          );
        });
    }
    setInitialValues(form, getName, {
      comment: analysisFormData.submission?.comment || analysisFormData.analysis.comment,
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const needToSelectSupervisor = () => {
    const role = findPractitionerRoleByOrganization(
      user.practitionerRoles,
      analysisFormData.proband?.organization_id!,
    );
    return isPractitionerResident(role!);
  };

  const onSearch = (searchText: string) => {
    if (searchText) {
      HybridApi.searchSupervisors({
        organizationId: analysisFormData.proband?.organization_id!,
        prefix: searchText,
      }).then(({ data }) => {
        setSupervisors(
          data?.supervisors.map((supervisor) => ({
            label: supervisor.license ? (
              <>
                {supervisor.name} &mdash; {supervisor.license}
              </>
            ) : (
              supervisor.name
            ),
            value: supervisor.id,
          })) ?? [],
        );
        setAutoCompleteDropdownIsOpen(true);
      });
    } else {
      setAutoCompleteDropdownIsOpen(false);
    }
  };

  return (
    <>
      <AnalysisForm
        form={form}
        className={styles.submissionForm}
        name={FORM_NAME}
        layout="vertical"
      >
        <div className={styles.supervisorCommentWrapper}>
          {needToSelectSupervisor() && (
            <Form.Item
              name={getName('resident_supervisor_id' satisfies keyof HybridAnalysis)}
              label={
                <ProLabel
                  title={intl.get('prescription.submission.responsable.doctor.label')}
                  colon
                />
              }
              rules={defaultFormItemsRules}
            >
              <Select
                showSearch
                placeholder={intl.get('prescription.submission.responsable.doctor.placeholder')}
                onSearch={onSearch}
                options={supervisors}
                onSelect={(value: string) => {
                  dispatch(
                    prescriptionFormActions.saveSubmissionStepData({
                      resident_supervisor_id: value,
                    }),
                  );
                }}
                open={autoCompleteDropdownIsOpen}
                onBlur={() => setAutoCompleteDropdownIsOpen(false)}
                defaultActiveFirstOption={false}
                showArrow={false}
                filterOption={false}
              />
            </Form.Item>
          )}
          <Form.Item
            name={getName('comment' satisfies keyof HybridAnalysis)}
            label={<ProLabel title={intl.get('prescription.submission.general.comment')} colon />}
          >
            <Input.TextArea rows={3} />
          </Form.Item>
        </div>
      </AnalysisForm>
      <ProLabel
        className={styles.reviewLabel}
        title={intl.get('prescription.add.parent.submission.verify.info.title')}
      />
      <Collapse
        {...defaultCollapseProps}
        bordered
        defaultActiveKey={['analyse', ...(config?.steps.map(({ title }) => title) ?? [])]}
      >
        <CollapsePanel key="analyse" header={intl.get('prescription.submission.title')}>
          <Descriptions className="label-20" column={1} size="small">
            <Descriptions.Item label={intl.get('prescription.submission.item.analysis.title')}>
              <Tag color="geekblue">{getAnalysisNameByCode(analysisType!, true)}</Tag>
            </Descriptions.Item>
            <Descriptions.Item label={intl.get('prescription.submission.item.prescribing.org')}>
              {analysisFormData.proband?.organization_id}
            </Descriptions.Item>
          </Descriptions>
        </CollapsePanel>
        {config?.steps
          .filter(({ id }) => id !== currentStep?.id)
          .map((step) => (
            <CollapsePanel
              key={step.title}
              header={step.title}
              extra={
                <FormOutlined
                  onClick={(event) => {
                    event.stopPropagation();
                    dispatch(
                      prescriptionFormActions.goTo({
                        index: step.index!,
                        lastStepIsNext: true,
                      }),
                    );
                  }}
                />
              }
            >
              {SubmissionStepMapping[step.id]}
            </CollapsePanel>
          ))}
      </Collapse>
    </>
  );
};

export default Submission;
