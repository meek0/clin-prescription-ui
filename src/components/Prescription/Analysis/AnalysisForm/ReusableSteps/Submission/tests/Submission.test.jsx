import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import Submission from '../index';

// stub to fix ANTD error
window.matchMedia =
  window.matchMedia ||
  function () {
    return {
      matches: false,
      addListener: function () {},
      removeListener: function () {},
    };
  };

describe('Submission', () => {
  test('should be as the snapshot', () => {
    const initialState = {
      global: {
        lang: null,
        analysisCodeMapping: {},
      },
      user: { user: { practitionerRoles: [] } },
      prescription: { analysisFormData: { analysis: {} } },
    };
    const store = configureStore()(initialState);
    const component = renderer.create(
      <Provider store={store}>
        <Submission />
      </Provider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchObject({});
    expect(tree).toMatchSnapshot();
  });
});
