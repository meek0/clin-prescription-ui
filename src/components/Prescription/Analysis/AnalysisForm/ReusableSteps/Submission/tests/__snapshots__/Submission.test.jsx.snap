// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submission should be as the snapshot 1`] = `
[
  <form
    className="ant-form ant-form-vertical"
    id="submission"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div>
      <div
        className="ant-form-item"
      >
        <div
          className="ant-row ant-form-item-row"
          style={{}}
        >
          <div
            className="ant-col ant-form-item-label ant-form-item-label-wrap"
            style={{}}
          >
            <label
              className=""
              htmlFor="submission_submission_comment"
              title=""
            >
              <span
                className="ant-typography"
                style={
                  {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                <span>
                  <span
                    className=""
                  >
                    Commentaire général
                  </span>
                  <span>
                    :
                  </span>
                </span>
              </span>
            </label>
          </div>
          <div
            className="ant-col ant-form-item-control"
            style={{}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <textarea
                  className="ant-input"
                  disabled={false}
                  id="submission_submission_comment"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onCompositionEnd={[Function]}
                  onCompositionStart={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  rows={3}
                  style={{}}
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>,
  <span
    className="ant-typography"
    style={
      {
        "WebkitLineClamp": undefined,
      }
    }
  >
    <span>
      <span
        className=""
      >
        Veuillez vérifier les informations ci-dessous avant de soumettre le formulaire.
      </span>
    </span>
  </span>,
  <div
    className="ant-collapse ant-collapse-icon-position-start"
    role={null}
  >
    <div
      className="ant-collapse-item ant-collapse-item-active"
    >
      <div
        aria-disabled={false}
        aria-expanded={true}
        className="ant-collapse-header"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="ant-collapse-expand-icon"
          onClick={null}
        >
          <span
            aria-label="caret-down"
            className="anticon anticon-caret-down ant-collapse-arrow"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="caret-down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
              />
            </svg>
          </span>
        </div>
        <span
          className="ant-collapse-header-text"
          onClick={null}
        >
          Analyse
        </span>
      </div>
      <div
        className="ant-collapse-content ant-collapse-content-active"
        role={null}
      >
        <div
          className="ant-collapse-content-box"
        >
          <div
            className="ant-descriptions ant-descriptions-small label-20"
          >
            <div
              className="ant-descriptions-view"
            >
              <table>
                <tbody>
                  <tr
                    className="ant-descriptions-row"
                  >
                    <td
                      className="ant-descriptions-item"
                      colSpan={1}
                    >
                      <div
                        className="ant-descriptions-item-container"
                      >
                        <span
                          className="ant-descriptions-item-label"
                          style={{}}
                        >
                          Analyse demandée
                        </span>
                        <span
                          className="ant-descriptions-item-content"
                          style={{}}
                        >
                          <span
                            className="ant-tag ant-tag-geekblue"
                            style={
                              {
                                "backgroundColor": undefined,
                              }
                            }
                          />
                        </span>
                      </div>
                    </td>
                  </tr>
                  <tr
                    className="ant-descriptions-row"
                  >
                    <td
                      className="ant-descriptions-item"
                      colSpan={1}
                    >
                      <div
                        className="ant-descriptions-item-container"
                      >
                        <span
                          className="ant-descriptions-item-label"
                          style={{}}
                        >
                          Établissement prescripteur
                        </span>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;